### ISSUE TITLE
[Docs] Deployment Help

### ISSUE THREAD

**Issue by Luc1412** [Rapptz/discord.py#5851]
## The Problem

Many people start creating a Discord Bot with maybe already good Python knowledge, but stuck at deploying there bot on a good way. I myself encountered the same issue which took me several days searching for bugs in code without knowing that it was a deployment issue. 
There already some useful information provided by the R. Danny's Tag, but why they should be hidden in the bot and not accessible to all people, because everyone who start creating a public bot will encounter these issues.

## The Ideal Solution

I wouldn't go too much in detail, but providing at least the base idea, tools and resources you need.
- What's required to get started at the beginning with smaller bots
- Scaling up your bot.  
- indicators when it's the right time to act

## The Current Solution

None

## Summary

Providing some basic help to deploy and scale your bot.

**Comment by drewburr** [comment]
@Luc1412 it sounds like you have some good information to share about this topic. Would you be able to provide additional detail to help get this started? Maybe some examples of issues you ran into? I feel that this is more likely to get off the ground if we have a good starting point.

**Comment by Luc1412** [comment]
I currently work on/learn a more professional deployment method (docker+kubernetes), but am not that experienced with those methods yet.

I would include following information:

- I wrote my first (eg. Hello World) bot and got my Linux VPS. How to get started.
-> mby. suggest using Systemd including some sample config
Maybe there are also some better deployment methods when you get started, but I choose systemd.

- Maybe some rough recommendations for VPS resources depending on the bot's size. (RAM, Cpu Cores, Networking, (unlimited) Traffic)

- indicators for lack of VPS resources/deployment method
-> some month ago I experienced random disconnects. I thought it was caused by my Network Connection or Discord itself
After a lot of research I came on the idea that's Python itself. Python is only single Threaded, so after a specific point it makes sense to use Cluster

- Professional deployment of your Discord Bot
-> Using Docker and eg. Kubernetes (sample Dockerfile+compose), Communication between cluster eg. using Redis

**Comment by ahnaf-zamil** [comment]
It is actually a good idea to show the developers about how they can deploy their bots. Because, people can make bots since d.py is well documented, but it is not shown how you can actually deploy it. 

In my opinion, the best way to deploy a bot (which would be mentioned in the docs) is to use Docker. This would work for both small and big bots since Docker is highly scalable and it's completely containerized. So in the docs, we could just tell them how they can deploy the bot step-by-step using a VPS with Docker on it.

I would include the following stuff:
  - Installing the dependencies on their VPS (python is pre-installed on all Linux distros)
  - Installing Docker, creating a container with the bot and running it 24/7 (it can be done easily with `-d` when running the container).
  
This is just my opinion about it. I would always assume the fact that the person who is reading the docs has no prior knowledge about deploying a bot, so it's better to go through everything step-by-step.

**Comment by Luc1412** [comment]
I agree that Docker is a great way to deploy and scale a bot, but it's already very advanced and causes a lot of issues with networking, volume management, image creating, etc. 
It could be mentioned as "advanced deployment", but I think systemd is the way to go for the average 100-5000 server bot.

**Comment by ahnaf-zamil** [comment]
@Luc1412 By `systemd`, do you mean creating a service in the OS to keep the bot running 24/7? If so, it's not a bad idea.

Let me share something that happened with me. When I was trying to deploy my first bot on AWS EC2, almost every deployment was showing how to use `systemd`. But I was having a lot of issues with `systemd` since I didn't know how it worked (back then, I was nub at stuff). So the bot wasn't running and I was frustrated. But the method that worked for **me**, was to use a Node executable called `pm2` which is basically a type of "easy-to-use" process manager and kept my bot running 24/7. I would just go to my bot's directory and run `pm2 start bot.py --name=mybot --interpreter python3`.

But I do agree that Docker can be a bit intimidating for newbies so it's fair to put it in the "advanced deployment" section.

**Comment by alkatar21** [comment]
I have 2 small 1 Guild bots, one using Docker and one systemd. Since I had used both before, it was not a big deal, even if I often had problems with reconnects in the beginning, although I don't know if that was due to Discord, because it became less by itself. But to have examples of this I would also find useful.
